backend/api/views/compute_plan_graph.py

```diff
@@ -18,121 +21,61 @@
-        {"id": "out/trunk_model", "kind": "model"},
-    ],
-    ComputeTask.Category.TASK_AGGREGATE: [{"id": "out/model", "kind": "model"}],
-    ComputeTask.Category.TASK_PREDICT: [{"id": "out/predictions", "kind": "model"}],
-}
-
 MAX_TASKS_DISPLAYED = 1000
 
 
-def _get_task_inputs(category):
-    return TASK_CATEGORY_INPUTS.get(category, [])
-
-
-def _get_task_outputs(category):
-    return TASK_CATEGORY_OUTPUTS.get(category, [])
-
-
-def _get_target_input(edge):
-    target_category = edge.get("target_task_category")
-    inputs = [input["id"] for input in TASK_CATEGORY_INPUTS.get(target_category, [])]
-    if TASK_CATEGORY_INPUTS.get(target_category) is None:
-        raise Exception(
-            (
-                "Failed to build CP workflow graph. Could not build an edge for: "
-                f"task {edge['target_task_key']} has a category {target_category}."
-            )
-        )
-    if (
-        target_category == ComputeTask.Category.TASK_COMPOSITE
-        and edge.get("source_task_category") == ComputeTask.Category.TASK_AGGREGATE
-    ):
-        input = inputs[1]
-    else:
-        input = inputs[0]
-
-    return input
-
-
-def _get_source_output(edge):
-    source_category = edge.get("source_task_category")
-    outputs = [output["id"] for output in TASK_CATEGORY_OUTPUTS.get(source_category, [])]
-    if TASK_CATEGORY_OUTPUTS.get(source_category) is None:
-        raise Exception(
-            (
-                "Failed to build CP workflow graph. Could not build an edge for: "
-                f"task {edge['target_task_key']} has a category {source_category}."
-            )
-        )
-    if (
-        source_category == ComputeTask.Category.TASK_COMPOSITE
-        and edge.get("target_task_category") == ComputeTask.Category.TASK_AGGREGATE
-    ):
-        output = outputs[1]
-    else:
-        output = outputs[0]
-
-    return output
+class JsonbBuildObj(Func):
+    function = "jsonb_build_object"
 
 
 @api_view(["GET"])
 def get_cp_graph(request, compute_plan_pk):
     """Return a workflow graph for each task of the computeplan"""
     validate_key(compute_plan_pk)
 
-    tasks = ComputeTask.objects.filter(compute_plan__key=compute_plan_pk, channel=get_channel_name(request)).values(
-        "key",
-        "rank",
-        "worker",
-        "status",
-        "category",
+    outputs = (
+        ComputeTask.objects.filter(key=OuterRef("pk"))
+        .annotate(
+            outputs_specs=JSONBAgg(
+                JsonbBuildObj(Value("identifier"), F("algo__outputs__identifier"), Value("kind"), "algo__outputs__kind")
+            ),
+        )
+        .values("outputs_specs")
+    )
+
+    inputs = (
+        ComputeTask.objects.filter(key=OuterRef("pk"))
+        .annotate(
+            inputs_specs=JSONBAgg(
+                JsonbBuildObj(Value("identifier"), F("algo__inputs__identifier"), Value("kind"), "algo__inputs__kind")
+            ),
+        )
+        .values("inputs_specs")
+    )
```

Comment:
Could you explain why you choose to serialize data in the query instead of using nested serializers (using `select_related` or `prefetch_related` in the query)?

Reply:
Here we would need 3 new serializer (to avoid serialize useless data, that would come from existing serializers), to build a non-standardized read-only view. The code was simpler and shorter doing it directly in the query, and as it is not a view that return proper task, input or output objects per se I feel serializers does not really bring added-value, for this particular case. 
But it's not a strong opinion, and it's still sometimes fuzzy to me when to absolutely use serializers and when it's not necessary in Django.

Reply:
After discussion, the fact that this view cannot be paginated yet and will manipulate a high volume of data, we stick with no serialization, for performance optimization. Comment added to explain the choice in the view.
Thanks Serge!